### PR TITLE
Registry based config - Part 1

### PR DESCRIPTION
**Pull request by dakinggg**
Adds the core of the registry system

Along for the ride, adds some __all__s. Will do a follow up PR that goes through and makes sure all the __all__s are correct.

- [x] Manual test all register styles (`registry-4-ia776l`, temporarily commented out some loggers and registered them through the different registry methods)
- [x] Write stub of the docs section
- [x] Pass through code again

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -259,6 +259,10 @@ export HUGGING_FACE_HUB_TOKEN=your-auth-token
 
 and uncomment the line containing `--hf_repo_for_upload ...` in the above call to `inference/convert_composer_to_hf.py`.
 
+# UNDER CONSTRUCTION: Registry
```
Comment (mvpatel2000) [inline README.md:262]: put the construction sign emoji please

File: llmfoundry/registry.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+from composer.optim import ComposerScheduler
+from torch.optim import Optimizer
+
+from llmfoundry.interfaces import CallbackWithConfig
+from llmfoundry.utils.registry_utils import create
+
+_loggers_description = """The loggers registry is used to register classes that implement the LoggerDestination interface.
+These classes are used to log data from the training loop, and will be passed to the loggers arg of the Trainer. The loggers
+will be constructed by directly passing along the specified kwargs to the constructor."""
+loggers = create('llmfoundry',
```
Comment (mvpatel2000) [inline llmfoundry/registry.py:16]: Is there no nicer way of doing this...?

File: scripts/train/train.py [hunk 3]
```diff
@@ -150,6 +151,18 @@ def build_composer_model(model_cfg: DictConfig,
...
+    code_paths = pop_config(cfg,
+                            'code_paths',
+                            must_exist=False,
+                            default_value=[],
+                            convert=True)
+    for code_path in code_paths:
+        try:
+            import_file(code_path)
+        except Exception as e:
+            raise ValueError(f'Error importing code from {code_path}: {e}')
```
Comment (mvpatel2000) [inline scripts/train/train.py:164]: This adds custom registry stuff?

File: REGISTRY.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# LLM Foundry Registry
```
Comment (mvpatel2000) [inline REGISTRY.md:1]: why separate file instead of README?

**Review by mvpatel2000 — COMMENTED** [review]
while we're at it, we should add trailing comma lint @b-chu ?

Mostly looks good to me. not doing detailed pass yet, but I like the design and no major comments. some minor questions mostly for me

**Comment by mvpatel2000** [comment]
What's the appropriate level for testing here?

File: llmfoundry/registry.py (hunk 2 above)
Comment (dakinggg) [reply llmfoundry/registry.py:16]: of doing what specifically?

File: scripts/train/train.py (hunk 3 above)
Comment (dakinggg) [reply scripts/train/train.py:164]: yeah, it allows you to inject your code that registers stuff

File: REGISTRY.md (hunk 4 above)
Comment (dakinggg) [reply REGISTRY.md:1]: the readme gets very long without indirection to sections that might get large. not a super strong opinion.

**Comment by dakinggg** [comment]
@mvpatel2000 I did manual test of all the registration methods (see description), and have unit tests for all the registering methods. What other testing are you thinking? I could run the regression tests against this PR, not sure how much that adds though.

File: llmfoundry/interfaces/callback_with_config.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+__all__ = ['CallbackWithConfig']
+
+
+class CallbackWithConfig(Callback, abc.ABC):
+
+    def __init__(self, config: dict[str, Any], *args: Any,
+                 **kwargs: Any) -> None:
+        del config, args, kwargs
+        pass
```
Comment (mvpatel2000) [inline llmfoundry/interfaces/callback_with_config.py:17]: Why is this necessary?

**Review by mvpatel2000 — COMMENTED** [review]
LGTM but would like more eyes on this given big PR. 

Do we need to update YAMLs?

File: REGISTRY.md (hunk 4 above)
Comment (mvpatel2000) [reply REGISTRY.md:1]: Ok, I think that's fair for now though imo it's a little annoying (I like preview on github)

**Comment by dakinggg** [comment]
@mvpatel2000 should not require any yaml updates as the yaml API is not changing.

File: llmfoundry/interfaces/callback_with_config.py (hunk 5 above)
Comment (dakinggg) [reply llmfoundry/interfaces/callback_with_config.py:17]: We have some callbacks that just take kwargs, and some callbacks that also need access to the full train config passed in. This is just to define the interface explicitly instead of implicitly. Let me add a docstring.

File: llmfoundry/utils/registry_utils.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    Raises:
+        ValueError: If the validation functions failed or the registered item is invalid
+
+    Returns:
+        Any: The constructed item from the registry
+    """
+    if kwargs is None:
+        kwargs = {}
+
+    registered_item = registry.get(name)
```
[the remark this answers is not in the document]
Comment (irenedea) [reply llmfoundry/utils/registry_utils.py:104]: or maybe `entry_point`?

File: llmfoundry/utils/registry_utils.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        return super().get_all()
+
+    def get_entry_point(self, name: str, default: Optional[T] = None) -> T:
+        return super().get_entry_point(name, default=default)
+
+    def get_entry_points(self) -> Dict[str, T]:
+        return super().get_entry_points()
+
+
+def create(
```
Comment (irenedea) [inline llmfoundry/utils/registry_utils.py:50]: how about `create_registry`?

Or you can make it static method so the usage is `TypedRegistry.create(...)`

File: REGISTRY.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+```yaml
+...
+code_paths:
+  - foundry_imports.py
+...
+```
+
+
+## Discovering registrable components
+Coming soon
```
Comment (irenedea) [inline REGISTRY.md:84]: can we already do this?

**Review by irenedea — APPROVED** [review]

File: tests/utils/test_builders.py [hunk 9]
```diff
@@ -62,7 +61,7 @@ def test_build_callback_fails():
 
 @pytest.mark.parametrize(
     'interval_key,interval_value',
-    [('interval', '10ba'), ('batch_log_interval', 10)],
+    [('interval', '10ba')],
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply tests/utils/test_builders.py:64]: it was around for backwards compat and was deprecated for a while (although never had a versioned deprecation warning), so im removing now

File: llmfoundry/utils/registry_utils.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    """
+    if catalogue.check_exists(*namespace):
+        raise catalogue.RegistryError(f'Namespace already exists: {namespace}')
+
+    return TypedRegistry[generic_type](namespace,
+                                       entry_points=entry_points,
+                                       description=description)
+
+
+def builder(
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/registry_utils.py:75]: I liked the short naming here because its scoped inside the `registry_utils` file. will decide after resopnding to other comments.

File: llmfoundry/utils/registry_utils.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+import catalogue
+
+__all__ = ['TypedRegistry', 'create', 'builder']
+
+T = TypeVar('T')
+S = TypeVar('S')
+
+
+class TypedRegistry(catalogue.Registry, Generic[T]):
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/registry_utils.py:20]: I think this is very generic and so reasonable to keep generically named.

File: REGISTRY.md (hunk 8 above)
Comment (dakinggg) [reply REGISTRY.md:84]: No, this is a follow on PR with the cli

File: scripts/train/train.py [hunk 12]
```diff
@@ -151,6 +151,15 @@ def build_composer_model(model_cfg: DictConfig,
 
 
 def main(cfg: DictConfig) -> Trainer:
+    # Run user provided code if specified
+    code_paths = pop_config(cfg,
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply scripts/train/train.py:155]: oops, bad merge

File: llmfoundry/utils/registry_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/registry_utils.py:75]: renamed to `construct_from_registry`

File: llmfoundry/utils/registry_utils.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+    def get_entry_point(self, name: str, default: Optional[T] = None) -> T:
+        return super().get_entry_point(name, default=default)
+
+    def get_entry_points(self) -> Dict[str, T]:
+        return super().get_entry_points()
+
+
+def create(
+    *namespace: str,
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/registry_utils.py:51]: the constructor takes sequence of str, so this lets you pass that in without making it a list or tuple yourself. all other args must be passed by kwarg

File: llmfoundry/utils/registry_utils.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        raise catalogue.RegistryError(f'Namespace already exists: {namespace}')
+
+    return TypedRegistry[generic_type](namespace,
+                                       entry_points=entry_points,
+                                       description=description)
+
+
+def builder(
+    name: str,
+    registry: catalogue.Registry,
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/registry_utils.py:77]: yeah, it should be, will fix

File: llmfoundry/utils/registry_utils.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    post_validation_function: Optional[Callable[[Any], None]] = None,
+    kwargs: Optional[Dict[str, Any]] = None,
+) -> Any:
+    """Helper function to build an item from the registry.
+
+    Args:
+        name (str): The name of the registered item
+        registry (catalogue.Registry): The registry to fetch the item from
+        partial_function (bool, optional): Whether to return a partial function for registered callables. Defaults to True.
+        pre_validation_function (Optional[Union[Callable[[Any], None], type]], optional): An optional validation function called
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/registry_utils.py:90]: `Optional` means that it can be `None`. `optional` means that it has a default and is not a required arg

[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/registry_utils.py:90]: added the explanation that it should throw

File: llmfoundry/utils/registry_utils.py (hunk 7 above)
Comment (dakinggg) [reply llmfoundry/utils/registry_utils.py:50]: yup, changed to `create_registry`

File: llmfoundry/utils/registry_utils.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (irenedea) [reply llmfoundry/utils/registry_utils.py:92]: ohhh haha

File: llmfoundry/utils/registry_utils.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (irenedea) [reply llmfoundry/utils/registry_utils.py:53]: Do we expect to create the registry in multiple namespaces?

**Review by mvpatel2000 — APPROVED** [review]

File: llmfoundry/utils/registry_utils.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/registry_utils.py:53]: see registry.py for how they're constructed, but basically that arg is like "llmfoundry", "loggers". not super important, could've been a single string like "llmfoundry.loggers". just a way of breaking up the hierarchy.
